content/blog/open-sourcing-workshops-on-demand-part3-understanding-the-backend.md

```diff
@@ -0,0 +1,352 @@
+{{ SCRIPTDIR }}/reset-$ws.sh
+{{ SCRIPTDIR }}/create-appliance.sh
+
+NAME=mllab
+TMPDIR=/tmp/$NAME.$stdid
+
+
+mkdir -p $TMPDIR
+
+# Define local variables
+echo wid=$wid
+APPMIN=`get_range_min $wid`
+echo stdid=$stdid
+echo APPMIN=$APPMIN
+mlport=$(($stdid-$APPMIN+{{ hostvars[inventory_hostname]['MLPORT-WKSHP-ML101'] }}))
+mlport2=$(($stdid-$APPMIN+{{ hostvars[inventory_hostname]['MLPORT2-WKSHP-ML101'] }}))
+httpport=$(($stdid-$APPMIN+{{ hostvars[inventory_hostname]['HTTPPORT-WKSHP-ML101'] }}))
+
+cat > $TMPDIR/dockerd-entrypoint.sh << EOF
+export HTTPPORT
+tini -g -- start-notebook.sh &
+sleep 3
+jupyter lab list | tail -1 | cut -d'=' -f2 | cut -d' ' -f1 > {{ STUDDIR }}/student$stdid/mltoken
+sleep infinity
+EOF
+
+cat > $TMPDIR/Dockerfile << EOF
+FROM ${NAME}:latest
+USER root
+COPY dockerd-entrypoint.sh /usr/local/bin/
+ENTRYPOINT /usr/local/bin/dockerd-entrypoint.sh
+RUN mkdir -p {{ STUDDIR }}/student$stdid
+RUN useradd student$stdid -u $stdid -g 100 -d {{ STUDDIR }}/student$stdid
+RUN chown student$stdid:users {{ STUDDIR }}/student$stdid
+# Unlock the account
+RUN perl -pi -e "s|^student$stdid:!:|student$stdid:\$6\$rl1WNGdr\$qHyKDW/prwoj5qQckWh13UH3uE9Sp7w43jPzUI9mEV6Y1gZ3MbDDMUX/1sP7ZRnItnGgBEklmsD8vAKgMszkY.:|" /etc/shadow
+# In case we need sudo
+#RUN echo "student$stdid   ALL=(ALL)       NOPASSWD: ALL" >> /etc/sudoers
+WORKDIR {{ STUDDIR }}/student$stdid
+USER student$stdid
+ENV NB_USER student$stdid
+ENV NB_UID $stdid
+ENV HTTPPORT $httpport
+RUN git clone https://github.com/snowch/ml-101 {{ STUDDIR }}/student$stdid/
+RUN /opt/conda/bin/jupyter-nbconvert --clear-output --inplace {{ STUDDIR }}/student$stdid/*.ipynb
+EOF
+
+
+# Look at https://stackoverflow.com/questions/34264348/docker-inside-docker-container
+# and http://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/
+# For security consider using https://github.com/nestybox/sysbox
+cat > $TMPDIR/docker-compose.yml << EOF
+version: '3.5'
+services:
+  $NAME$stdid:
+    image: $NAME$stdid
+    build: .
+    #privileged: true
+    ports:
+      - "$httpport:8888"
+      - "$mlport:4040"
+      - "$mlport2:4041"
+#    volumes:
+#      - /var/run/docker.sock:/var/run/docker.sock
+EOF
+cat > $TMPDIR/launch-$NAME << EOF
+#!/bin/bash
+cd $TMPDIR
+docker-compose up --build -d
+EOF
+```
+
+1﻿3- T﻿he `copy_folder yml` playbook is now executed to deploy the notebooks and scripts necessary for the participant to run the workshop. Remember that the  participant got a student allocated to him at the time of the registration. This student is picked from a range that is allocated for the workshop. The admin decides on the maximum capacity it allocates to a given workshop. `c﻿opy_folder.yml`: This is historically one of very first playbook we used and therefore a very important one. It performs the necessary actions to deploy, personnalize (by substituting ansible variables) the selected notebook to the appropriate student home folder.
```

Comment:
student can be a man or a women.
Remember that the participant got a student allocated to him/her....
Is it "student" or "student Id" ?